### ISSUE TITLE
ReferenceSearchResult should optionally return the access type (read/write)

### ISSUE THREAD

**Issue by kjeremy** [rust-analyzer/rust-analyzer#2560]
`textDocument/documentHghlight` allows us to fill in a kind of read and write to say: "this text range represents a read" for example and can highlight differently based on this flag. I can also imagine filtering reference results by read or write access. We should return this information in `ReferenceSearchResult`.

**Comment by kjeremy** [comment]
@matklad 

How do I tell if a `NameRef` matches the lhs of an assignment operator (basically... is this a write to that location)? Is the correct way to test the name ref's range with the `ast::BinExpr`'s `lhs` range?

**Comment by matklad** [comment]
Good question... A funny case here is something like this:

```rust
x[y] = 92
```

Here, `x` is write but `y` is read (or at least one would expect that).

I *think* it makes sense to just start with simplistic syntactic analysis, which only handles `foo =`. Long term, I think we need exactly the notion of access paths that borrowck uses? Ie:

```rust
let mut x = 92;
let r = &mut x; // this is a write access, despite being on rhs
*r = 92
```
